BoogieLang/Ast_Cfg_Transformation.thy

```diff
@@ -0,0 +1,2280 @@
+
+    from this \<open>binop_eval_val Ge v1 v2 = (Some (BoolV False))\<close> have "binop_greaterOrEqual lit1 lit2 = Some (LBool False)" by simp 
+
+    from this obtain i1 i2 where
+      "lit1 = LInt i1" and
+      "lit2 = LInt i2" by (metis binop_greaterOrEqual.elims option.simps(3))
+
+    from this \<open>binop_greaterOrEqual lit1 lit2 = Some (LBool False)\<close> have "\<not>(i1 \<ge> i2)" by simp
+    hence "i2 > i1" by simp
+    hence "binop_greater lit2 lit1 = (Some (LBool True))" by (simp add: \<open>lit1 = LInt i1\<close> \<open>lit2 = LInt i2\<close>)
+    hence "binop_eval Gt lit2 lit1 = Some (LBool True)" by (simp add: \<open>lit1 = LInt i1\<close> \<open>lit2 = LInt i2\<close>)
+    hence "binop_eval_val Gt (LitV lit2) (LitV lit1) = Some (BoolV True)" by simp 
+    thus ?thesis using neg_ge redE1 redE2  \<open>v1 = (LitV lit1)\<close> \<open>v2 = (LitV lit2)\<close> True by (simp add: RedBinOp) 
+  next
+    case False
+    from this assms neg_ge have 
+      "A,\<Lambda>,\<Gamma>,\<Omega> \<turnstile> \<langle>UnOp unop.Not (e1 \<guillemotleft>Ge\<guillemotright> e2), ns\<rangle> \<Down> BoolV False" by simp
+    hence eq_false: "A,\<Lambda>,\<Gamma>,\<Omega> \<turnstile> \<langle>(e1 \<guillemotleft>Ge\<guillemotright> e2), ns\<rangle> \<Down> BoolV True" by (metis not_false_equals_true)
+    
+    from this obtain v1 v2 where
+      redE1: "A,\<Lambda>,\<Gamma>,\<Omega> \<turnstile> \<langle>e1, ns\<rangle> \<Down> v1" and
+      redE2: "A,\<Lambda>,\<Gamma>,\<Omega> \<turnstile> \<langle>e2, ns\<rangle> \<Down> v2" by blast
+
+    from this eq_false have "binop_eval_val Ge v1 v2 = (Some (BoolV True))" by (metis RedBinOp_case expr_eval_determ(1))
+
+    from this obtain lit1 lit2 where
+      "v1 = (LitV lit1)" and 
+      "v2 = (LitV lit2)" by (metis binop_eval_val.simps(32) binop_eval_val.simps(33) option.discI val.exhaust)
+
+    from this \<open>binop_eval_val Ge v1 v2 = (Some (BoolV True))\<close> have "binop_greaterOrEqual lit1 lit2 = Some (LBool True)" by simp 
+
+    from this obtain i1 i2 where
+      "lit1 = LInt i1" and
+      "lit2 = LInt i2" by (metis binop_greaterOrEqual.elims option.simps(3))
+
+    from this \<open>binop_greaterOrEqual lit1 lit2 = Some (LBool True)\<close> have "(i1 \<ge> i2)" by simp
+    hence "\<not>(i2 > i1)" by simp
+    hence "binop_greater lit2 lit1 = (Some (LBool False))" by (simp add: \<open>lit1 = LInt i1\<close> \<open>lit2 = LInt i2\<close>)
+    hence "binop_eval Gt lit2 lit1 = Some (LBool False)" by (simp add: \<open>lit1 = LInt i1\<close> \<open>lit2 = LInt i2\<close>)
+    hence "binop_eval_val Gt (LitV lit2) (LitV lit1) = Some (BoolV False)" by simp 
+    thus ?thesis using neg_ge redE1 redE2  \<open>v1 = (LitV lit1)\<close> \<open>v2 = (LitV lit2)\<close> False by (simp add: RedBinOp) 
+  qed
+qed
+
+(* TODO: Can I avoid needing this? *)
+fun inv_into_assertion :: "expr \<Rightarrow> cmd" where
+  "inv_into_assertion e = (Assert e)"
```

Comment:
This function is the same as just `Assert`. One should replace all occurrences of `inv_into_assertion` by `Assert`.